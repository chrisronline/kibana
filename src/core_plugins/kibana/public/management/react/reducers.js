/* eslint-disable */
import { combineReducers } from 'redux';
import { routerReducer } from 'react-router-redux'

import app from './app/app.reducer';
import indexPattern from './index-pattern/index-pattern.reducer';

export const rootReducer = combineReducers({
  routing: routerReducer,
  app,
  indexPattern,
});
