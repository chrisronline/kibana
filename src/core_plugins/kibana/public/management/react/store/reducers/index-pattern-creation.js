import { handleActions } from 'redux-actions';
import { createSelector } from 'reselect';
import { chunk, sortBy as sortByLodash } from 'lodash';

import {
  selectTimeField,
  fetchedTimeFields,
  goToNextPage,
  goToPreviousPage,
  changeSort,
  fetchedIndices,
  includeSystemIndices,
  excludeSystemIndices,
  creatingIndexPattern,
  createdIndexPattern,
} from '../actions/index-pattern-creation';

import {
  getIndexPatternCreate,
} from '../../reducers';

const defaultState = {
  isIncludingSystemIndices: false,
  isCreating: false,
  timeFields: {
    timeFields: undefined,
    selectedTimeField: undefined,
  },
  results: {
    indices: undefined,
    pattern: undefined,
    hasExactMatches: false,
    perPage: 10,
    page: 1,
    sortBy: undefined,
    sortAsc: true,
  }
}

export default handleActions({
  [selectTimeField](state, { payload }) {
    const { timeFields } = payload;

    return {
      ...state,
      timeFields: {
        ...state.timeFields,
        timeFields,
      },
    };
  },
  [fetchedTimeFields](state, { payload }) {
    const { timeFields } = payload;

    return {
      ...state,
      timeFields: {
        ...state.timeFields,
        timeFields,
      },
    };
  },
  [selectTimeField](state, { payload }) {
    const { timeField } = payload;

    return {
      ...state,
      timeFields: {
        ...state.timeFields,
        selectedTimeField: timeField,
      },
    };
  },
  [fetchedIndices](state, { payload }) {
    const { indices, hasExactMatches, pattern } = payload;

    return {
      ...state,
      results: {
        ...state.results,
        indices,
        hasExactMatches,
        pattern,
      },
    };
  },
  [changeSort](state, { payload }) {
    const { sortBy, sortAsc } = payload;

    return {
      ...state,
      results: {
        ...state.results,
        sortBy,
        sortAsc: sortBy === state.results.sortBy ? !state.results.sortAsc: sortAsc,
      },
    };
  },
  [goToNextPage](state, { payload }) {
    return {
      ...state,
      results: {
        ...state.results,
        page: getNextPage(state.results),
      },
    };
  },
  [goToPreviousPage](state, { payload }) {
    return {
      ...state,
      results: {
        ...state.results,
        page: getPreviousPage(state.results),
      },
    };
  },
  [includeSystemIndices](state, action) {
    return {
      ...state,
      isIncludingSystemIndices: true,
    };
  },
  [excludeSystemIndices](state, action) {
    return {
      ...state,
      isIncludingSystemIndices: false,
    };
  },
  [creatingIndexPattern](state, action) {
    return {
      ...state,
      isCreating: true,
    };
  },
  [createdIndexPattern](state, action) {
    return {
      ...state,
      isCreating: false,
    };
  },
}, defaultState);

function getNextPage({ page, perPage, indices }) {
  const pages = Math.ceil(indices.length / perPage);
  const nextPage = page + 1;
  return nextPage > pages
    ? 1
    : nextPage;
}

function getPreviousPage({ page, perPage, indices }) {
  const pages = Math.ceil(indices.length / perPage);
  const previousPage = page - 1;
  return previousPage < 1
    ? pages
    : previousPage;
}

function getFilteredIndices(indices, isIncludingSystemIndices) {
  if (!indices) {
    return indices;
  }

  if (isIncludingSystemIndices) {
    return indices;
  }

  // All system indices begin with a period.
  return indices.filter(index => !index.name.startsWith('.'));
}

function getPaginatedIndices(indices, page, perPage) {
  const pagesOfIndices = chunk(indices, perPage);
  return pagesOfIndices[page - 1];
}

export const getFilteredAndPaginatedIndices = createSelector(
  [
    state => state.results,
    state => state.isIncludingSystemIndices
  ],
  ({
    indices: allIndices,
    page,
    perPage,
    sortBy,
    sortAsc,
  }, isIncludingSystemIndices) => {
    if (allIndices === undefined) {
      return {
        indices: undefined,
        numOfPages: 0,
      };
    }

    let filteredIndices = getFilteredIndices(allIndices, isIncludingSystemIndices);
    if (!!sortBy) {
      filteredIndices = sortByLodash(filteredIndices, sortBy);
      if (!sortAsc) {
        filteredIndices.reverse();
      }
    }

    const numOfPages = Math.ceil(filteredIndices.length / perPage);
    const indices = getPaginatedIndices(filteredIndices, page, perPage);
    return {
      indices,
      numOfPages,
    }
  }
);

export const getPattern = state => getIndexPatternCreate(state).results.pattern;
export const getSelectedTimeField = state => getIndexPatternCreate(state).timeFields.selectedTimeField;
