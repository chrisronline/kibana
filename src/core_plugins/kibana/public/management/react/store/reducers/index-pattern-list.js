import { handleActions } from 'redux-actions';

import {
  fetchedIndexPatterns,
  setTransientTableId,
} from '../actions/index-pattern-list';

import {
  getIndexPatternList,
} from '../../reducers';

const defaultState = {
  indexPatterns: undefined,
  transientTableId: undefined,
};

export default handleActions({
  [fetchedIndexPatterns](state, { payload }) {
    const { indexPatterns } = payload;

    return {
      ...state,
      indexPatterns,
    };
  },
  [setTransientTableId](state, { payload: { id: transientTableId } }) {
    return {
      ...state,
      transientTableId,
    };
  }
}, defaultState);

export const getIndexPatterns = state => getIndexPatternList(state).indexPatterns;
export const getPathToIndexPatterns = () => 'indexPatterns';
