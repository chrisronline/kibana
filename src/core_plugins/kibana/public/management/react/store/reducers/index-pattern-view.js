import { handleActions } from 'redux-actions';

import {
  fetchedIndexPattern,
  setAsDefaultIndexPattern,
} from '../actions/index-pattern-view';

import {
  getIndexPatternView,
} from '../../reducers';

const defaultState = {
  indexPattern: {
    pattern: undefined,
    fields: undefined,
    timeFieldName: undefined,
    isDefault: undefined,
    id: undefined,
  },
  fieldsTable: {},
  tabs: {},
};

export const indexPatternView = handleActions({
  [fetchedIndexPattern](state, { payload: { indexPattern } }) {
    return {
      ...state,
      indexPattern,
    };
  },
  [setAsDefaultIndexPattern](state) {
    return {
      ...state,
      indexPattern: {
        ...state.indexPattern,
        isDefault: true,
      }
    };
  }
}, defaultState);

export const getPathToFields = () => 'indexPattern.fields';
export const getTabs = state => getIndexPatternView(state).tabs;
