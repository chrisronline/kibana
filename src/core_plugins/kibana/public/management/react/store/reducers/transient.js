import { handleActions } from 'redux-actions';

import {
  change,
  createItem,
} from '../actions/transient';

const defaultState = {};

export default handleActions({
  [createItem](state, { payload: { id, defaultData }}) {
    return {
      ...state,
      [id]: defaultData || {},
    }
  },
  [change](state, { payload: { id, data } }) {
    return {
      ...state,
      [id]: {
        ...state[id],
        ...data
      },
    };
  },
}, defaultState);
