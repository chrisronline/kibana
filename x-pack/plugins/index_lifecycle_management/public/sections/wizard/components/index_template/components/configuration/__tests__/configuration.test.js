/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import { Configuration } from '../';
import { indexLifecycleManagement } from '../../../../../../../store/reducers';

jest.mock('@elastic/eui', () => ({
  EuiDescribedFormGroup: props => props.children || '',
  EuiFlexGroup: props => props.children || '',
  EuiFlexItem: props => props.children || '',
  EuiSpacer: props => props.children || '',
  EuiSelect: props => props.children || '',
  EuiFieldNumber: props => props.children || '',
  EuiCallOut: props => props.children || '',
  EuiButtonEmpty: props => props.children || '',
  EuiLink: props => props.children || '',
  EuiDescribedFormGroup: props => props.children || '',
}));

jest.mock('../../../../../../../store/actions', () => ({
  setSelectedNodeAttrs: jest.fn(),
  fetchNodes: jest.fn(),
  setSelectedPrimaryShardCount: jest.fn(),
  setSelectedReplicaCount: jest.fn()
}));
jest.mock('../../../../../form_errors', () => ({
  ErrableFormRow: props => props.children || '',
}));
jest.mock('../../../../node_attrs_details', () => ({
  NodeAttrsDetails: props => props.children || '',
}));

const defaultStore = {
  subscribe: jest.fn(),
  dispatch: jest.fn(),
};

describe('Configuration', () => {
  it('should render properly', () => {
    const defaultState = indexLifecycleManagement({}, {});
    const store = {
      getState: () => ({
        ...defaultState,
        indexTemplate: {
          ...defaultState.indexTemplate,
          indexTemplates: [{}]
        }
      }),
      ...defaultStore,
    };

    const component = mount(
      <Provider store={store}>
        <Configuration
          validate={() => {}}
          isShowingErrors={false}
          errors={{}}
        />
      </Provider>
    );

    expect(component.find('Configuration')).toMatchSnapshot();
  });

  it('should show a warning about shard counts if they are higher than node attrs count', () => {
    const defaultState = indexLifecycleManagement({}, {});
    const store = {
      getState: () => ({
        ...defaultState,
        indexTemplate: {
          ...defaultState.indexTemplate,
          indexTemplates: [{}]
        },
        nodes: {
          ...defaultState.nodes,
          selectedPrimaryShardCount: 2,
          selectedNodeAttrs: 'foo',
          nodes: {
            foo: [1]
          }
        }
      }),
      ...defaultStore,
    };

    const component = mount(
      <Provider store={store}>
        <Configuration
          validate={() => {}}
          isShowingErrors={false}
          errors={{}}
        />
      </Provider>
    );

    expect(component.find('EuiCallOut').get(1)).toMatchSnapshot();
  });

  describe('should validate fields when they change', () => {
    const defaultState = indexLifecycleManagement({}, {});
    const actions = require('../../../../../../../store/actions');
    const store = {
      getState: () => ({
        ...defaultState,
        indexTemplate: {
          ...defaultState.indexTemplate,
          indexTemplates: [{}]
        },
      }),
      ...defaultStore,
    };

    test('for node attrs', async () => {
      const validate = jest.fn();
      const component = mount(
        <Provider store={store}>
          <Configuration
            validate={validate}
            isShowingErrors={false}
            errors={{}}
          />
        </Provider>
      );

      await component.find('EuiSelect').prop('onChange')({ target: { value: '1' } });
      expect(validate).toHaveBeenCalled();
      expect(actions.setSelectedNodeAttrs).toHaveBeenCalledWith('1');
    });

    test('for primary shards', async () => {
      const validate = jest.fn();
      const component = mount(
        <Provider store={store}>
          <Configuration
            validate={validate}
            isShowingErrors={false}
            errors={{}}
          />
        </Provider>
      );

      await component.find('EuiFieldNumber').get(0).props.onChange({ target: { value: '1' } });
      expect(validate).toHaveBeenCalled();
      expect(actions.setSelectedPrimaryShardCount).toHaveBeenCalledWith('1');
    });

    test('for replicas', async () => {
      const validate = jest.fn();
      const component = mount(
        <Provider store={store}>
          <Configuration
            validate={validate}
            isShowingErrors={false}
            errors={{}}
          />
        </Provider>
      );

      await component.find('EuiFieldNumber').get(1).props.onChange({ target: { value: '1' } });
      expect(validate).toHaveBeenCalled();
      expect(actions.setSelectedReplicaCount).toHaveBeenCalledWith('1');
    });
  });

  it('should show the flyout', () => {
    const defaultState = indexLifecycleManagement({}, {});
    const store = {
      getState: () => ({
        ...defaultState,
        indexTemplate: {
          ...defaultState.indexTemplate,
          indexTemplates: [{}]
        }
      }),
      ...defaultStore,
    };

    const component = mount(
      <Provider store={store}>
        <Configuration
          validate={() => {}}
          isShowingErrors={false}
          errors={{}}
        />
      </Provider>
    );

    component.find('Configuration').instance().setState({ isShowingNodeDetailsFlyout: true });
    component.update();

    expect(component.find('NodeAttrsDetails')).toMatchSnapshot();
  });
});
