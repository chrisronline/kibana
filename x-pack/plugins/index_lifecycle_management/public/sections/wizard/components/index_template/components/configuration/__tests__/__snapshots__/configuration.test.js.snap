// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Configuration should render properly 1`] = `
<Configuration
  errors={Object {}}
  fetchNodes={[Function]}
  isPrimaryShardCountHigherThanSelectedNodeAttrsCount={false}
  isShowingErrors={false}
  nodeOptions={
    Array [
      Object {
        "text": "-- Do not reallocate my indices --",
        "value": undefined,
      },
    ]
  }
  selectedNodeAttrs=""
  selectedPrimaryShardCount={1}
  selectedReplicaCount={1}
  setSelectedNodeAttrs={[Function]}
  setSelectedPrimaryShardCount={[Function]}
  setSelectedReplicaCount={[Function]}
  validate={[Function]}
>
  <div>
    <EuiDescribedFormGroup
      description="A hot index is actively being written to."
      fullWidth={true}
      title={
        <h4>
          Configure hot indices
        </h4>
      }
      titleSize="s"
    >
      <ErrableFormRow
        errorKey="node_attrs"
        errors={Object {}}
        helpText={null}
        isShowingErrors={false}
        label="Where do you want your hot indices to live?"
      >
        <EuiSelect
          onChange={[Function]}
          options={
            Array [
              Object {
                "text": "-- Do not reallocate my indices --",
                "value": undefined,
              },
            ]
          }
          value=""
        />
      </ErrableFormRow>
      <EuiCallOut
        color="warning"
        size="s"
        title={
          <p>
            The best way to determine how many shards you need is to benchmark using realistic data and queries on your hardware.
             
            <EuiLink
              href="https://www.elastic.co/webinars/using-rally-to-get-your-elasticsearch-cluster-size-right"
            >
              Learn more
            </EuiLink>
          </p>
        }
      />
      <EuiSpacer />
      <EuiFlexGroup>
        <EuiFlexItem
          style={
            Object {
              "maxWidth": 188,
            }
          }
        >
          <ErrableFormRow
            errorKey="primary_nodes"
            errors={Object {}}
            isShowingErrors={false}
            label="Primary shards"
          >
            <EuiFieldNumber
              onChange={[Function]}
              value={1}
            />
          </ErrableFormRow>
        </EuiFlexItem>
        <EuiFlexItem
          style={
            Object {
              "maxWidth": 188,
            }
          }
        >
          <ErrableFormRow
            errorKey="replicas"
            errors={Object {}}
            isShowingErrors={false}
            label="Replicas"
          >
            <EuiFieldNumber
              onChange={[Function]}
              value={1}
            />
          </ErrableFormRow>
        </EuiFlexItem>
      </EuiFlexGroup>
    </EuiDescribedFormGroup>
  </div>
</Configuration>
`;

exports[`Configuration should show a warning about shard counts if they are higher than node attrs count 1`] = `
<EuiCallOut
  color="warning"
  iconType="help"
  style={
    Object {
      "marginTop": "1rem",
    }
  }
  title="Primary shard count is too high"
>
  The shard count should be lower than the number of nodes that match the selected attributes.
</EuiCallOut>
`;

exports[`Configuration should show the flyout 1`] = `
<NodeAttrsDetails
  close={[Function]}
  selectedNodeAttrs=""
/>
`;
